distributor.go

```diff
@@ -176,53 +184,87 @@ func (d *Distributor) Append(ctx context.Context, samples []*model.Sample) error
 
 	d.receivedSamples.Add(float64(len(samples)))
 
-	samplesByIngester := map[string][]*model.Sample{}
-	for _, sample := range samples {
-		key := tokenForMetric(userID, sample.Metric)
-		ingesters, err := d.cfg.Ring.Get(key, d.cfg.ReplicationFactor)
-		if err != nil {
-			return err
+	sampleTrackers := make([]sampleTracker, len(samples), len(samples))
+	samplesByIngester := map[string][]*sampleTracker{}
+	keys := make([]uint32, len(samples), len(samples))
+
+	for i, sample := range samples {
+		keys[i] = tokenForMetric(userID, sample.Metric)
+	}
+
+	ingesters, err := d.cfg.Ring.BatchGet(keys, d.cfg.ReplicationFactor, ring.Write)
+	if err != nil {
+		return err
+	}
+
+	for i := range samples {
+		sampleTrackers[i] = sampleTracker{
+			sample:     samples[i],
+			minSuccess: (len(ingesters[i]) / 2) + 1,
+			succeeded:  0,
+		}
+
+		// Skip those that have not heartbeated in a while. NB these are still
+		// included in the calculation of minSuccess, so if too many failed ingesters
+		// will cause the whole write to fail.
+		liveIngesters := make([]string, len(ingesters[i]))
+		for _, ingester := range ingesters[i] {
+			if time.Now().Sub(ingester.Timestamp) <= d.cfg.HeartbeatTimeout {
+				liveIngesters = append(liveIngesters, ingester.Hostname)
+			}
+		}
+		if len(liveIngesters) < sampleTrackers[i].minSuccess {
+			return fmt.Errorf("wanted at least %d live ingesters to process write, had %d",
+				sampleTrackers[i].minSuccess, len(liveIngesters))
```

Comment:
Do we really want to stop writing all of the samples because one sample failed to write to a sufficient number of replicas?